routes/utils.js

```diff
@@ -1,7 +1,113 @@
 var request = require("request");
+var async = require("async");
+var url = require("url");
+var NodeFormData = require("form-data");
 var Path = require("path");
+
 var Constants = require("../constants");
 
+function createProject(config, user, data, callback) {
+  var project = JSON.parse(JSON.stringify(data));
+  var createURL = config.publishURL + "/projects";
+  delete project.id;
+
+  if(!user) {
+    callback(null, 200, project);
+    return;
+  }
+
+  request({
+    method: "POST",
+    uri: createURL,
+    headers: {
+      "Authorization": "token " + user.token
+    },
+    body: project,
+    json: true
+  }, function(err, response, body) {
+    if(err) {
+      console.error("Failed to send request to " + createURL + " with: ", err);
+      callback(null, 500);
+      return;
+    }
+
+    if(response.statusCode !== 201) {
+      callback(response.body, response.statusCode);
+      return;
+    }
+
+    callback(null, 200, body);
+  });
+}
+
+function persistProjectFiles(config, user, project, data, callback) {
+  var publishURL = config.publishURL + "/files";
+  var files = {};
+
+  function persist(file, callback) {
+    var options = url.parse(publishURL);
+    options.method = "POST";
+    options.headers = { "Authorization": "token " + user.token };
+    var formData = new NodeFormData();
+    formData.append("path", file.path);
+    formData.append("project_id", project.id);
+    if(file.stream) {
+      formData.append("buffer", file.stream, { knownLength: file.size });
+    } else {
+      formData.append("buffer", file.buffer, { filename: file.path });
+    }
+
+    formData.submit(options, function(err, response) {
+      var body = "";
+
+      if(err) {
+        console.error("Failed to send request to " + publishURL + " with: ", err);
+        callback({ status: 500 });
+        return;
+      }
+
+      response.on('error', function(err) {
+        console.error("Failed to receive response from " + publishURL + " with: ", err);
+        callback({ status: 500 });
+      });
+
+      response.on('data', function(data) {
+        body += data;
+      });
+
+      response.on('end', function() {
+        body = JSON.parse(body);
+
+        if(response.statusCode !== 201) {
+          callback({ message: body, status: response.statusCode });
+          return;
+        }
+
+        files[body.path] = {
+          id: body.id,
+          project_id: body.project_id,
+          path: body.path
+        };
+
+        callback();
+      });
+    });
+  }
+
+  if(!user) {
+    callback(null, 200);
```

Comment:
I'd prefer to see the conditional in the control logic to make the different code paths more clear. It would also get rid of the weird stubbing of callbacks